File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/modland/RemoteCatalog.java
/**
 * @file
 * @brief Remote implementation of catalog
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs.modland;

import android.net.Uri;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.io.IOException;
import java.io.InputStream;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import app.zxtune.Log;
import app.zxtune.fs.HtmlUtils;
import app.zxtune.fs.api.Cdn;
import app.zxtune.fs.http.MultisourceHttpProvider;
import app.zxtune.utils.ProgressCallback;
import app.zxtune.utils.StubProgressCallback;

/**
 * There's additional api gate for xbmc plugin at
 * http://www.exotica.org.uk/mediawiki/extensions/ExoticASearch/Modland_xbmc.php
 * but it seems to be not working and has no such wide catalogue as http gate does.
 */
public class RemoteCatalog implements Catalog {

  private static final String TAG = RemoteCatalog.class.getName();

  private static final Pattern PAGINATOR =
          Pattern.compile("Browsing (.+?) - (\\d+) results? - showing page (\\d+) of (\\d+).");

  private final MultisourceHttpProvider http;
  private final Grouping authors;
  private final Grouping collections;
  private final Grouping formats;

  RemoteCatalog(MultisourceHttpProvider http) {
    this.http = http;
    this.authors = new BaseGrouping("aut", "Modules from author ");
    this.collections = new BaseGrouping("col", "Modules from collection ");
    this.formats = new BaseGrouping("for", "Modules in format ");
  }

  @Override
  public Grouping getAuthors() {
    return authors;
  }

  @Override
  public Grouping getCollections() {
    return collections;
  }

  @Override
  public Grouping getFormats() {
    return formats;
  }

  private static Uri.Builder getMainUriBuilder() {
    return new Uri.Builder().scheme("https").authority("www.exotica.org.uk").path("mediawiki" +
        "/index.php").appendQueryParameter("title", "Special:Modland");
  }

  private static Uri.Builder getGroupsUriBuilder(String type, String filter) {
    return getMainUriBuilder().appendQueryParameter("md", "b_" + type).appendQueryParameter("st",
        filter);
  }

  private static Uri.Builder getTracksUriBuilder(String type, int id) {
    return getMainUriBuilder().appendQueryParameter("md", type).appendQueryParameter("id",
        Integer.toString(id));
  }

  private class BaseGrouping implements Grouping {

    private final String tag;
    private final String headerPrefix;

    BaseGrouping(String tag, String headerPrefix) {
      this.tag = tag;
      this.headerPrefix = headerPrefix;
    }

    @Override
    public void queryGroups(String filter, final GroupsVisitor visitor,
                            final ProgressCallback progress) throws IOException {
      Log.d(TAG, "queryGroups(type=%s, filter=%s)", tag, filter);
      loadPages(getGroupsUriBuilder(tag, filter), new PagesVisitor() {

        private int total = 0;
        private int done = 0;

        @Override
        public boolean onPage(String header, int results, Document doc) {
          if (total == 0) {
            total = results;
            visitor.setCountHint(total);
          }
          done += parseGroups(doc, visitor);
          progress.onProgressUpdate(done, total);
          return true;
        }
      });
    }

    @Override
    public Group getGroup(final int id) throws IOException {
      Log.d(TAG, "getGroup(type=%s, id=%d)", tag, id);
      final Group[] resultRef = new Group[1];
      loadPages(getTracksUriBuilder(tag, id), (header, results, doc) -> {
        if (header.startsWith(headerPrefix)) {
          final String name = header.substring(headerPrefix.length());
          resultRef[0] = new Group(id, name, results);
        }
        return false;
      });
      final Group result = resultRef[0];
      if (result != null) {
        return result;
      }
      throw new IOException(String.format(Locale.US,"Failed to find group with id=%d", id));
    }

    @Override
    public void queryTracks(int id, final TracksVisitor visitor, final ProgressCallback progress) throws IOException {
      Log.d(TAG, "queryGroupTracks(type=%s, id=%d)", tag, id);
      loadPages(getTracksUriBuilder(tag, id), new PagesVisitor() {

        private int total = 0;
        private int done = 0;

        @Override
        public boolean onPage(String header, int results, Document doc) {
          if (total == 0) {
            total = results;
            visitor.setCountHint(total);
          }
          final int tracks = parseTracks(doc, visitor);
          done += tracks;
          progress.onProgressUpdate(done, total);
          return tracks != 0;
        }
      });
    }

    @Override
    public Track getTrack(int id, final String filename) throws IOException {
      Log.d(TAG, "getGroupTrack(type=%s, id=%d, filename=%s)", tag, id, filename);
      final Track[] resultRef = {null};
      queryTracks(id, new TracksVisitor() {
        // TODO: remove
        @Override
        public void setCountHint(int size) {}

        @Override
        public boolean accept(Track obj) {
          if (obj.filename.equals(filename)) {
            resultRef[0] = obj;
            return false;
          } else {
            return true;
          }
        }
      }, StubProgressCallback.instance());
      final Track result = resultRef[0];
      if (result != null) {
        return result;
      }
      throw new IOException(String.format(Locale.US, "Failed to get track '%s' with id=%d", filename, id));
    }
  }

  private static int parseGroups(Document doc, GroupsVisitor visitor) {
    int result = 0;
    for (Element el : doc.select("table:has(>caption)>tbody>tr:has(>td>a[href*=md=])")) {
      final Element nameEl = el.child(0).child(0);
      final Element countEl = el.child(1);
      final Integer id = HtmlUtils.getQueryInt(nameEl, "id");
      final String name = nameEl.text();
      final Integer tracks = HtmlUtils.tryGetInteger(countEl.text());
      if (id != null && name != null && tracks != null) {
        visitor.accept(new Group(id, name, tracks));
        ++result;
      }
    }
    return result;
  }

  private static int parseTracks(Document doc, TracksVisitor visitor) {
    int result = 0;
    for (Element el : doc.select("table:has(>caption)>tbody>tr:has(>td>a[href*=file=pub/])")) {
      final Element pathEl = el.child(0).child(0);
      final Element sizeEl = el.child(4);
      final String href = pathEl.attr("href");
      final int pathPos = href.indexOf("pub/modules/");
      final String path = "/" + href.substring(pathPos);
      final Integer size = HtmlUtils.tryGetInteger(sizeEl.text());
      if (size != null) {
        if (!visitor.accept(new Track(path, size))) {
          return 0;
        }
        ++result;
      }
    }
    return result;
  }

  public static Uri[] getTrackUris(String id) {
    return new Uri[]{
        Cdn.modland(id),
        Uri.parse("http://ftp.amigascne.org/mirrors/ftp.modland.com" + id)
    };
  }

  interface PagesVisitor {
    boolean onPage(String title, int results, Document doc);
  }

  private void loadPages(Uri.Builder query, PagesVisitor visitor) throws IOException {
    for (int pg = 1; ; ++pg) {
      final Document doc = HtmlUtils.parseDoc(readPage(query, pg));
      final Element hdr = doc.selectFirst("table>caption");
      final Matcher matcher = hdr != null ? PAGINATOR.matcher(hdr.text()) : null;
      if (matcher != null && matcher.find()) {
        Log.d(TAG, "Load page: " + hdr.text());
        final String header = matcher.group(1);
        final String results = matcher.group(2);
        final String page = matcher.group(3);
        final String pagesTotal = matcher.group(4);
        if (pg != Integer.parseInt(page)) {
          throw new IOException("Invalid paginator structure");
        }
        if (visitor.onPage(header, Integer.parseInt(results), doc)
                && pg < Integer.parseInt(pagesTotal)) {
          continue;
        }
      }
      break;
    }
  }

  private InputStream readPage(Uri.Builder base, int start) throws IOException {
    final Uri uri = base.appendQueryParameter("pg", Integer.toString(start)).build();
    return http.getInputStream(uri);
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/modland/RemoteCatalog.kt
/**
 * @file
 * @brief Remote implementation of catalog
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs.modland

import android.net.Uri
import app.zxtune.Logger
import app.zxtune.fs.HtmlUtils
import app.zxtune.fs.Jsoup.findFirstText
import app.zxtune.fs.api.Cdn
import app.zxtune.fs.api.Proxy
import app.zxtune.fs.http.MultisourceHttpProvider
import app.zxtune.utils.ProgressCallback
import app.zxtune.utils.StubProgressCallback
import org.jsoup.nodes.Document
import java.io.IOException
import java.util.regex.Pattern
import androidx.core.net.toUri

/**
 * There's additional api gate for xbmc plugin at
 * http://www.exotica.org.uk/mediawiki/extensions/ExoticASearch/Modland_xbmc.php
 * but it seems to be not working and has no such wide catalogue as http gate does.
 */

private val LOG = Logger(RemoteCatalog::class.java.name)
private val PAGINATOR: Pattern =
    Pattern.compile("Browsing (.+?) - (\\d+) results? - showing page (\\d+) of (\\d+).")

open class RemoteCatalog(val http: MultisourceHttpProvider) : Catalog {

    private val authors by lazy { BaseGrouping("aut", "Modules from author ") }
    private val collections by lazy { BaseGrouping("col", "Modules from collection ") }
    private val formats by lazy { BaseGrouping("for", "Modules in format ") }

    override fun getAuthors(): Catalog.Grouping = authors

    override fun getCollections(): Catalog.Grouping = collections

    override fun getFormats(): Catalog.Grouping = formats

    private inner class BaseGrouping(val tag: String, val headerPrefix: String) : Catalog.Grouping {
        override fun queryGroups(
            filter: String, visitor: Catalog.Visitor<Group>, progress: ProgressCallback
        ) {
            LOG.d { "queryGroups(type=${tag}, filter=${filter})" }
            var total = 0
            var done = 0
            loadPages(getGroupsUri(tag, filter)) { _, results, doc ->
                if (total == 0) {
                    total = results
                }
                done += parseGroups(doc, visitor)
                progress.onProgressUpdate(done, total)
                true
            }
        }

        override fun getGroup(id: Int): Group {
            LOG.d { "getGroup(type=${tag}, id=${id})" }
            val resultRef = arrayOfNulls<Group>(1)
            loadPages(getTracksUri(tag, id)) { header, results, _ ->
                if (header.startsWith(headerPrefix)) {
                    val name = header.substring(headerPrefix.length)
                    resultRef[0] = Group(id, name, results)
                }
                false
            }
            return resultRef[0] ?: throw IOException("Failed to find group with id=${id}")
        }

        override fun queryTracks(
            id: Int, visitor: Catalog.Visitor<Track>, progress: ProgressCallback
        ) {
            LOG.d { "queryGroupTracks(type=${tag}, id=${id})" }
            var total = 0
            var done = 0
            loadPages(getTracksUri(tag, id)) { _, results, doc ->
                if (total == 0) {
                    total = results
                }
                val tracks = parseTracks(doc, visitor)
                done += tracks
                progress.onProgressUpdate(done, total)
                tracks != 0
            }
        }

        override fun getTrack(id: Int, filename: String): Track {
            LOG.d { "getGroupTrack(type=${tag}, id=${id}, filename=${filename})" }
            var result: Track? = null
            queryTracks(id, {
                if (it.filename == filename) {
                    result = it
                }
            }, StubProgressCallback.instance())
            return result ?: throw IOException("Failed to get track '${filename}' with id=${id}")
        }
    }

    companion object {

        @JvmStatic
        fun getTrackUris(id: String) = arrayOf(
            Cdn.modland(id), "https://ftp.amigascne.org/mirrors/ftp.modland.com$id".toUri()
        )
    }

    private fun loadPages(
        query: Uri, onPage: (title: String, results: Int, doc: Document) -> Boolean
    ) {
        for (pg in 1..Int.MAX_VALUE) {
            val doc = HtmlUtils.parseDoc(readPage(query, pg))
            val hdr = doc.findFirstText("table>caption") ?: break
            val matcher = PAGINATOR.matcher(hdr)
            if (!matcher.find()) break
            LOG.d { "Load page: $hdr" }
            val header = matcher.group(1) ?: break
            val results = matcher.group(2) ?: break
            val page = matcher.group(3) ?: break
            val pagesTotal = matcher.group(4) ?: break
            if (pg != page.toInt()) {
                throw IOException("Invalid paginator structure")
            }
            if (onPage(header, results.toInt(), doc) && pg < pagesTotal.toInt()) {
                continue
            }
            break
        }
    }

    private fun readPage(base: Uri, start: Int) =
        base.buildUpon().appendQueryParameter("pg", start.toString()).build().let { uri ->
            http.getInputStream(arrayOf(Proxy.uriFor(uri), uri))
        }
}

private fun getMainUriBuilder() =
    Uri.Builder().scheme("https").authority("www.exotica.org.uk").path("mediawiki/index.php")
        .appendQueryParameter("title", "Special:Modland")

private fun getGroupsUri(type: String, filter: String) =
    getMainUriBuilder().appendQueryParameter("md", "b_${type}").appendQueryParameter("st", filter)
        .build()

private fun getTracksUri(type: String, id: Int) =
    getMainUriBuilder().appendQueryParameter("md", type).appendQueryParameter("id", id.toString())
        .build()

private fun parseGroups(doc: Document, visitor: Catalog.Visitor<Group>) =
    doc.select("table:has(>caption)>tbody>tr:has(>td>a[href*=md=])").onEach { el ->
        val nameEl = el.child(0).child(0)
        val countEl = el.child(1)
        val id = HtmlUtils.getQueryInt(nameEl, "id") ?: return@onEach
        val name = nameEl.text()
        val tracks = HtmlUtils.tryGetInteger(countEl.text()) ?: return@onEach
        if (name.isNotEmpty()) {
            visitor.accept(Group(id, name, tracks))
        }
    }.size

private fun parseTracks(doc: Document, visitor: Catalog.Visitor<Track>) =
    doc.select("table:has(>caption)>tbody>tr:has(>td>a[href*=file=pub/])").onEach { el ->
        val pathEl = el.child(0).child(0)
        val sizeEl = el.child(4)
        val href = pathEl.attr("href")
        val pathPos = href.indexOf("pub/modules/")
        val path = "/${href.substring(pathPos)}"
        HtmlUtils.tryGetInteger(sizeEl.text())?.let { size ->
            visitor.accept(Track(path, size))
        }
    }.size
